extensions/azurePublishNew/src/node/provisionService.ts

```diff
@@ -0,0 +1,92 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+import { IBotProject } from '@botframework-composer/types';
+
+import { getAppRegistrationProvisionService } from './azureResources/appRegistration';
+import { getWebAppProvisionService } from './azureResources/webApp';
+import { getBotChannelProvisionService } from './azureResources/botChannel';
+import { getAzureFunctionsProvisionService } from './azureResources/azureFunctions';
+import { getCosmosDbProvisionService } from './azureResources/cosmosDb';
+import { getLuisAuthoringProvisionService } from './azureResources/luisAuthoring';
+import { getLuisPredictionProvisionService } from './azureResources/luisPrediction';
+import { getBlogStorageProvisionService } from './azureResources/blobStorage';
+import { getQnAProvisionService } from './azureResources/qna';
+import { getAppServiceProvisionService } from './azureResources/servicePlan';
+
+// bot project => candidate resources => select & configure resources => order & provision
+
+export type ProvisionWorkingSet = Record<string, object>;
+
+export type ResourceProvisionService = {
+  getDependencies: () => string[];
+  getRecommendationForProject: (project: IBotProject) => 'required' | 'optional' | 'invalid';
+  provision: <TConfig>(config: TConfig, workingSet: ProvisionWorkingSet) => void;
+};
+
+export type ResourceDefinition = {
+  key: string;
+  text: string;
+  description: string;
+  tier: string;
+  group: string;
+};
+
+export type GetResourcesResult = ResourceDefinition & {
+  required: boolean;
+};
+
+type ProvisionConfig = {
+  key: string;
+};
+
+type ProvisionCredentials = {
+  token: string;
+  graphToken: string;
+  subscriptionId: string;
+};
+
+export type ResourceConfig = {
+  key: string;
+};
+
+export type ProvisionMethod = <TConfig>(config: TConfig, workingSet: ProvisionWorkingSet) => ProvisionWorkingSet;
+
+export const getProvisionServices = (): Record<string, ResourceProvisionService> => {
+  return {
+    appRegistration: getAppRegistrationProvisionService(),
+    webApp: getWebAppProvisionService(),
+    botRegistration: getBotChannelProvisionService(),
+    azureFunctionApp: getAzureFunctionsProvisionService(),
+    cosmosDB: getCosmosDbProvisionService(),
+    appInsights: getWebAppProvisionService(),
+    luisAuthoring: getLuisAuthoringProvisionService(),
+    luisPrediction: getLuisPredictionProvisionService(),
+    blobStorage: getBlogStorageProvisionService(),
+    qna: getQnAProvisionService(),
+    servicePlan: getAppServiceProvisionService(),
+  };
+};
+
+export const setUpProvisionService = (creds: ProvisionCredentials) => {
```

Comment:
I can see two ways to get the token, graphToken to the services.  

1. You could pass it to each service in getProvisionServices and have the services hold it via closure.  This is good because you can't create services without tokens, you don't have to check if you have a token all the time, and the token doesn't affect every method.  The downside is that tokens expire and you have to re-create the services whenever there is a new token.
2. You could pass the credentials to each provision service method.  This reduces the chance that any service will be holding onto an expired token, but is super tedious to author and call the methods.

Reply:
going with approach 1